pkg/query-service/app/queryBuilder/functions.go

```diff
@@ -0,0 +1,286 @@
+package queryBuilder
+
+import (
+	"math"
+	"sort"
+
+	v3 "go.signoz.io/signoz/pkg/query-service/model/v3"
+)
+
+// funcCutOffMin cuts off values below the threshold and replaces them with NaN
+func funcCutOffMin(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value < threshold {
+				point.Value = math.NaN()
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcCutOffMax cuts off values above the threshold and replaces them with NaN
+func funcCutOffMax(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value > threshold {
+				point.Value = math.NaN()
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcClampMin cuts off values below the threshold and replaces them with the threshold
+func funcClampMin(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value < threshold {
+				point.Value = threshold
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcClampMax cuts off values above the threshold and replaces them with the threshold
+func funcClampMax(result *v3.Result, threshold float64) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			if point.Value > threshold {
+				point.Value = threshold
+			}
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcAbsolute returns the absolute value of each point
+func funcAbsolute(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			point.Value = math.Abs(point.Value)
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcLog2 returns the log2 of each point
+func funcLog2(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			point.Value = math.Log2(point.Value)
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcLog10 returns the log10 of each point
+func funcLog10(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		for idx, point := range series.Points {
+			point.Value = math.Log10(point.Value)
+			series.Points[idx] = point
+		}
+	}
+	return result
+}
+
+// funcCumSum returns the cumulative sum for each point in a series
+func funcCumSum(result *v3.Result) *v3.Result {
+	for _, series := range result.Series {
+		var sum float64
+		for idx, point := range series.Points {
+			if !math.IsNaN(point.Value) {
```

Comment:
should this check be part of all functions? Eg `funcClampMax`?

Reply:
No, it's only needed where the other values would do arithmetic and produce NaN result. The threshold functions will not need this as any comparison with NaN returns and false and operation is skipped https://go.dev/play/p/nb6aUSZBbUY

Reply:
so `sum += NaN` will evaluate to NaN?

Reply:
Yes, we skip NaNs in functions where there is some arithmetic.

Reply:
got it